JAX transformer text classification

mtsokol opened the pull request:
This PR adds a tutorial with JAX transformer text classification task on IMDB dataset. It follows one of the [Keras tutorials](https://keras.io/examples/nlp/text_classification_with_transformer/). It's still missing the narrative.

@vfdev-5 I took most of the training, validation and visualization code from https://github.com/jax-ml/jax-ai-stack/pull/95.

I perform dataset preprocessing in a separate function in order to show usage of `grain.RandomAccessDataSource`.

mtsokol commented on docs/JAX_transformer_text_classification.md:
````diff
@@ -0,0 +1,355 @@
+## Dataset
+
+```{code-cell} ipython3
+def prepare_imdb_dataset(num_words: int) -> tuple:
````
A few bits here are taken from https://github.com/keras-team/keras/blob/495a570c5350e689d686c572009811cc4492d8c6/keras/src/datasets/imdb.py#L13 to fully reproduce data setup of [a Keras tutoral](https://keras.io/examples/nlp/text_classification_with_transformer/). I will simplify it.

mtsokol replied:
I revisited dataset preprocessing. Done!

melissawm commented on docs/JAX_transformer_text_classification.md:
```diff
@@ -0,0 +1,375 @@
+to know if a review is overall positive or negative, so we're facing a classification task.
+
+For that purpose, we're going to build a model comprised of a transformer and dense layers.
+What is special about transformers is that they utilize attention mechanism that allows
+the model to focus on the most relevant parts of the input sequence when processing information.
```
```suggestion
What is special about transformers is that they utilize an attention mechanism that allows
the model to focus on the most relevant parts of the input sequence when processing information.
```

mtsokol replied:
Done!

melissawm commented on docs/JAX_transformer_text_classification.md:
````diff
@@ -0,0 +1,375 @@
+x_test = pad_sequences(x_test, max_len=maxlen)
+```
+
+As our dataset are relatively small NumPy arrays we define it as a `RandomAccessDataSource`:
````
Maybe link to the PyGrain docs?
```suggestion
As our dataset are relatively small NumPy arrays we define it as a `RandomAccessDataSource` (see [PyGrain Data Sources](https://github.com/google/grain/blob/main/docs/data_sources.md) for more details):
```

mtsokol replied:
Done!

melissawm commented on docs/JAX_transformer_text_classification.md:
````diff
@@ -0,0 +1,375 @@
+```
+
+## Training
+
````
Can we add a few words about using nnx.Optimizer and the other nnx objects for training here? I think it's worth highlighting

mtsokol replied:
Done! I added two sentences about training.

melissawm commented on docs/JAX_transformer_text_classification.md:
````diff
@@ -0,0 +1,375 @@
+axs[1].plot(eval_metrics_history["test_accuracy"])
+```
+
+As we can se we acquired over 85% accuracy after 10 epochs. From the loss plot
+we see that in the last few ones loss value didn't improve much.
````
```suggestion
As we can see, we acquired over 85% accuracy after 10 epochs. From the loss plot,
we see that in the last few epochs the loss value didn't improve much.
```

melissawm replied:
Does this mean we should have chosen less epochs?

mtsokol replied:
IMO it means we converged - I expanded this section a bit.

melissawm reviewed (commented):
Couple of notes:
- It may be worth adding a few words about the Grain usage.
- I would add some "Next steps" notes, either finalizing with an interpretation of the results (why does our model work? Could we have chosen another one? What was the advantage in using the jax-ai-stack here?)

I have another more holistic comment for @jakevdp : do we want to be more explicit in these tutorials about which tools we are using are part of the jax-ai-stack? As an external reader, it is not obvious to me and I think we could do a better job of highlighting why the jax ecosystem tools are helpful in the examples.

mtsokol commented:
> It may be worth adding a few words about the Grain usage.

Sure! I added a few sentences about Grain.

> I would add some "Next steps" notes, either finalizing with an interpretation of the results (why does our model work? Could we have chosen another one? What was the advantage in using the jax-ai-stack here?)

I expanded the last section a bit with what came to my mind - let me know if it should be expanded further.

> do we want to be more explicit in these tutorials about which tools we are using are part of the jax-ai-stack?

I added `Tools overview` section with a few links to the most relevant libraries.

melissawm reviewed (approved):
Looks great, thanks @mtsokol !

trallard commented:
@jakevdp this should be ready now 🚀

mtsokol commented on docs/JAX_transformer_text_classification.md:
```diff
@@ -0,0 +1,402 @@
+
+## Tools overview
+
+Here's a list of key packages that belong to JAX AI stack:
```
Done!

jakevdp commented on docs/JAX_transformer_text_classification.md:
````diff
@@ -0,0 +1,402 @@
+axs[0].plot(eval_metrics_history["test_loss"])
+axs[1].set_title("Accuracy on test set")
+axs[1].plot(eval_metrics_history["test_accuracy"])
+```
````
Could we somehow print a selection of (truncated) test samples and labels? Without that it's hard for the reader to tell what exactly the model is doing.

mtsokol replied:
Done! I added a cell at the end where I inspect a few decoded samples together with a predicted and an actual label.

I squashed the PR, it's ready from my side!

mtsokol commented on docs/JAX_transformer_text_classification.md:
````diff
@@ -0,0 +1,402 @@
+
+```{code-cell} ipython3
+model = MyModel(rngs=nnx.Rngs(0))
+nnx.display(model)
````
Done!

jakevdp reviewed (approved):
Thanks!
